17 - add applicant information section to application forms

caravinci opened the pull request:
As specified in issue #17.

Leveraging the validation introduced with the `introduction` section, this PR should mostly introduce additional content and their corresponding field schemas... plus a bit of Typescript love, as suggested by @anncatton, with the section state constants/enum.

![image](https://user-images.githubusercontent.com/2107110/120376281-b1bc2c80-c2e9-11eb-88b7-2c95873e24f8.png)


Thanks for reviewing!

caravinci commented on components/pages/Applications/ApplicationForm/Forms/constants.tsx:
```diff
@@ -31,3 +43,254 @@ export const sectionsOrder = [
   'signature',
 ] as const; // to infer these as union rather than string[]
+
+export const countriesList = [
```
constants for the `country` and (honorific) `title`s.
the countries may come from an API endpoint in the future, as per conversation with @blabadi @kcullion @rosibaj

caravinci commented on components/pages/Applications/ApplicationForm/Forms/index.tsx:
```diff
@@ -119,6 +120,33 @@ const ApplicationFormsBase = ({ appId = 'none' }): ReactElement => {
                 }
+
+                // for the horizontal design in this app
+                [class$='-FormControl'] {
```
uniform styles for all the `FormControl` component groups (to enable the label being aside the field, rather than on top, etc.)
Adds a basic level of responsiveness, which may need to be improved, but won't look 100% horrible in smaller viewports.

caravinci commented on components/pages/Applications/ApplicationForm/Forms/types.ts:
```diff
@@ -26,8 +36,23 @@ type FormSectionValidationState_SectionsGenericType<T extends Record<string, For
   FormSectionValidationState_SectionsGenericType<{}>;
-export type FormSectionValidationState_Applicant =
-  FormSectionValidationState_SectionsGenericType<{}>;
+export type FormSectionValidationState_Applicant = FormSectionValidationState_SectionsGenericType<{
```
the new fields' type

caravinci commented on components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts:
```diff
@@ -1,23 +1,56 @@
 
-export const handleFieldTypes = (event: ChangeEvent<HTMLInputElement>) => {
-  if (['INPUT'].includes(event?.target?.tagName)) {
+export const schemaValidator = (fieldSchema: any, value: any) =>
```
abstracting this function to be reusable between field/overall validation requests.

caravinci commented on components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts:
```diff
@@ -1,23 +1,56 @@
+  }));
+
+export const getFieldDataFromEvent: FormFieldDataFromEvent = (event) => {
+  if (['INPUT', 'MULTISELECT', 'SELECT'].includes(event?.target?.tagName)) {
```
handling new target/event types

anncatton replied:
Maybe also put these event types as an enum, so we have a set of accepted types? Unless the uikit form components considers this somewhere, i'm not totally familiar with that section

caravinci commented on components/pages/Applications/ApplicationForm/Forms/validations/index.ts:
```diff
@@ -52,58 +55,52 @@ export const validationReducer = (
+  (validationState, dispatch) =>
+  (origin, validateSection) =>
+  async (field, value, shouldPersistResults) => {
+    if (validateSection) {
```
improving the field/overall validation routine, the other one was a bit unwieldy to reason with, passing state through the request (rather than the state always been there, as in now line 58) and whatnot...

caravinci commented on components/pages/Applications/ApplicationForm/Forms/validations/index.ts:
```diff
@@ -52,58 +55,52 @@ export const validationReducer = (
+
+      return results;
+    } else if (field) {
+      const { error } = await schemaValidator(yup.reach(combinedSchema[origin], field), value);
```
using the abstracted+reusable function I've commented about earlier.
you pass it either an individual field's schema, or a group of them (as in the overall section validation).

caravinci commented on components/pages/Applications/ApplicationForm/Forms/validations/index.ts:
```diff
@@ -158,59 +155,74 @@ export const useLocalValidation = (
+          const canBlur = !(
+            ['text'].includes(fieldType) &&
+            ['address_country'].includes(field) &&
+            localState[field]?.value
```
this because `Multiselect` still has some iffy behaviours to be managed... didn't have enough time to figure the internals for the accessible aspects of it (using it with a keyboard, etc.)

caravinci commented on components/pages/Applications/ApplicationForm/Forms/validations/schemas.ts:
```diff
@@ -21,13 +23,59 @@ yup.setLocale({
   },
 });
 
+export const applicantSchema = yup.object().shape({
```
the new validation schema 🚓

caravinci commented on package.json:
```diff
@@ -12,7 +12,7 @@
     "@emotion/styled": "^10.0.11",
     "@icgc-argo/ego-token-utils": "^8.2.0",
-    "@icgc-argo/uikit": "^1.10.1",
+    "@icgc-argo/uikit": "^1.10.3",
```
the few changes to `Select` and `Multiselect` to cover some basic accessibility validation edge cases.
still work pending but this should be ok for now.

caravinci commented on components/pages/Applications/ApplicationForm/Forms/DoubleFieldRow.tsx:
```diff
@@ -0,0 +1,66 @@
+
+import FormFieldHelpBubble from './FormFieldHelpBubble';
+
+const DoubleFieldRow = ({
```
this new component to allow two fields in a same row, half-row long fields, or the little blue helper bubbles shown in the screen capture.

caravinci commented on components/pages/Applications/ApplicationForm/Forms/FormFieldHelpBubble.tsx:
```diff
@@ -0,0 +1,47 @@
+import Typography from '@icgc-argo/uikit/Typography';
+import { useTheme } from '@icgc-argo/uikit/ThemeProvider';
+
+const HelpBubble = ({ tail, text }: { tail?: 'left' | 'right'; text: string }): ReactElement => {
```
the little help bubble.
I'd have reused the UiKit Tooltip but that component needs a serious overhaul to work well.

caravinci commented on components/pages/Applications/ApplicationForm/Forms/constants.tsx:
```diff
@@ -1,9 +1,21 @@
+const Applicant = dynamic(() => import('./applicant'));
 const Introduction = dynamic(() => import('./introduction'));
 
+export const FormSectionOverallStates = {
```
Now, for `la piece de resistance`, the overall state constants.

anncatton replied:
i think this is ok, but i was thinking an actual enum 🙂 . Like:
`export enum FormSectionOverallStates { CAN_EDIT = 'canEdit', COMPLETE = 'complete', ... }`

caravinci commented on components/pages/Applications/ApplicationForm/Forms/types.ts:
```diff
@@ -1,19 +1,29 @@
-  | 'pristine';
+type FormSectionOverallObj = typeof FormSectionOverallStates;
+
+export type FormSectionOverallState = FormSectionOverallObj[keyof FormSectionOverallObj];
```
which are used here to generate the union type used everywhere else to validate those states passed by the dictionary/enum.

caravinci commented on components/pages/Applications/ApplicationForm/Forms/helpers.tsx:
```diff
@@ -1,32 +1,38 @@
-) => sections.filter((sectionName) => !(state[sectionName]?.overall === 'disabled'));
+) =>
+  sections.filter(
+    (sectionName) => !(state[sectionName]?.overall === FormSectionOverallStates.DISABLED),
```
example of those pretty new constants being used... 🎉

caravinci commented on components/pages/Applications/ApplicationForm/Forms/Outline/helpers.ts:
```diff
@@ -14,38 +13,33 @@ import { FormSectionOverallStates } from '../types';
       };
 
-    case 'mustEdit':
+    case FormSectionOverallStates.MUSTEDIT:
```
must!

anncatton replied:
snake case for the edit enums so more readable? as much as i love something that sounds like its musty 😛

caravinci commented on components/pages/Applications/ApplicationForm/Forms/Outline/Section.tsx:
```diff
@@ -41,7 +42,11 @@ const FormSection = ({
+        FormSectionOverallStates.DISABLED,
+        FormSectionOverallStates.PRISTINE,
+        FormSectionOverallStates.TOUCHED,
+      ].includes(status) && <ValidationIcon status={status} />}
```
much order for great good

anncatton replied:
🎉 🎉 🎉

caravinci reviewed (commented):
Should be a much simpler PR than the one before, but the now "usual" set of comments to help walk through it 👍

caravinci commented on components/pages/Applications/ApplicationForm/Forms/constants.tsx:
```diff
@@ -1,9 +1,21 @@
 import dynamic from 'next/dynamic';
-const Introduction = dynamic(() => import('./introduction'));
+const Applicant = dynamic(() => import('./Applicant'));
+const Introduction = dynamic(() => import('./Introduction'));
```
Also, just realised I had named the files incorrectly. changing it now, before the issue compounds later.

anncatton commented on components/pages/Applications/ApplicationForm/Forms/constants.tsx:
```diff
@@ -31,3 +43,254 @@ export const sectionsOrder = [
+  { code: 'ZW', name: 'Zimbabwe' },
+];
+
+export const honorificsList = ['Dr. ', 'Miss', 'Mr.', 'Mrs.', 'Ms.', 'Prof.'];
```
Does `'Dr.'` need a space after? (the others don't have this so just verifying it's not a typo)

caravinci replied:
mindless copy pasta. lol

anncatton reviewed (commented):
Just me harping on a bit more about enums, but otherwise looks good!

caravinci commented:
All very good points!👌
Let me see what I can do.

caravinci commented:
@anncatton thanks for the suggestions.
I've corrected the unnecessarily socially distanced Dr., made the form states enum (and renamed it to constant case), and made the event target tags an enum as well; although I don't really love this last one, to be honest, so I take ideas if you can think of a better way to do that one.

caravinci commented on components/pages/Applications/ApplicationForm/Forms/types.ts:
```diff
@@ -1,19 +1,39 @@
+}
+
+export enum FORM_STATES {
+  CANEDIT = 'canEdit',
```
totally!
